From: Huang Ying <ying.huang@intel.com>
Subject: ACPI, APEI, Add WHEA _OSC support
References: bnc#697859
Patch-Mainline: not yet, submitted for 3.1

Signed-off-by: Thomas Renninger <trenn@suse.de>

APEI firmware first mode must be turned on explicitly on some
machines, otherwise there may be no GHES hardware error record for
hardware error notification.  APEI bit in generic _OSC call can be
used to do that, but on some machine, a special WHEA _OSC call must be
used.  This patch adds the support to that WHEA _OSC call.

Signed-off-by: Huang Ying <ying.huang@intel.com>
---
 drivers/acpi/apei/ghes.c |   10 ++++++++++
 1 file changed, 10 insertions(+)

Index: linux-2.6.32-SLE11-SP2/drivers/acpi/apei/ghes.c
===================================================================
--- linux-2.6.32-SLE11-SP2.orig/drivers/acpi/apei/ghes.c
+++ linux-2.6.32-SLE11-SP2/drivers/acpi/apei/ghes.c
@@ -679,6 +679,16 @@ static int __init ghes_init(void)
 	if (rc)
 		goto err_ioremap_exit;
 
+	rc = apei_osc_setup();
+	if (rc == 0 && osc_sb_apei_support_acked)
+		pr_info(GHES_PFX "APEI firmware first mode is enabled by APEI bit and WHEA _OSC.\n");
+	else if (rc == 0 && !osc_sb_apei_support_acked)
+		pr_info(GHES_PFX "APEI firmware first mode is enabled by WHEA _OSC.\n");
+	else if (rc && osc_sb_apei_support_acked)
+		pr_info(GHES_PFX "APEI firmware first mode is enabled by APEI bit.\n");
+	else
+		pr_info(GHES_PFX "Failed to enable APEI firmware first mode.\n");
+
 	return 0;
 err_ioremap_exit:
 	ghes_ioremap_exit();
